File: mudlib/lib/serialize.lua
-- mudlib/lib/serialize.lua — Lua values back into Lua source.
--
-- What OLC writes. Every generated area file goes through here, so the property
-- that matters is not "does it look right" but:
--
--     load(emit(v))() == v          for every v this accepts
--     emit(load(emit(v))()) == emit(v)      byte for byte
--
-- The second is the one that catches things. `codegen_d` built its output by
-- concatenating strings — `'short = "' .. data.short .. '",'` — so a room whose
-- title contained a quote produced a file that would not compile, and its
-- multi-line branch indented the closing `]]`, which put four spaces inside the
-- string and grew them on every read-write cycle for ever.
--
-- ─── Why not lib/jsonsafe.lua ────────────────────────────────────────────────
--
-- It answers a different question with a different rule set. `jsonsafe` refuses
-- a table that is simultaneously a list and a map because *JSON* has no such
-- type — Lua does, and a mob's `loot_table` beside its `count` is ordinary here.
-- It also has nothing to say about float round-tripping, which is most of the
-- hard part below. And its contract is pinned to `lua_to_json` by
-- `tests/state_cache.rs`, so widening it to serve a second consumer would break
-- a pact that exists for a reason.
--
-- The breadcrumb helpers are duplicated from it rather than shared, so the two
-- produce error messages that read alike without either owning the other. Six
-- lines is cheaper than a third module that exists to hold six lines.
--
-- See docs/src/lua-api/olc.md.

local M = {}

--- How deep a value may nest. A pathological structure should error rather than
--- eat the C stack — the same reasoning as `jsonsafe.MAX_DEPTH`.
M.MAX_DEPTH = 32

--- A scalar array shorter than this renders on one line.
M.INLINE_WIDTH = 78

local KEYWORDS = {
    ["and"] = true, ["break"] = true, ["do"] = true, ["else"] = true,
    ["elseif"] = true, ["end"] = true, ["false"] = true, ["for"] = true,
    ["function"] = true, ["goto"] = true, ["if"] = true, ["in"] = true,
    ["local"] = true, ["nil"] = true, ["not"] = true, ["or"] = true,
    ["repeat"] = true, ["return"] = true, ["then"] = true, ["true"] = true,
    ["until"] = true, ["while"] = true,
}

--- A breadcrumb, spelled the way `jsonsafe` spells it.
local function join(path, key)
    if path == "" then return tostring(key) end
    if type(key) == "number" then return path .. "[" .. tostring(key) .. "]" end
    return path .. "." .. tostring(key)
end

local function finite(n)
    return n == n and n ~= math.huge and n ~= -math.huge
end

-- ─── Scalars ─────────────────────────────────────────────────────────────────

--- Can this string be a bare table key?
---
--- Keywords are excluded as well as non-identifiers: `end = 1` is a syntax
--- error, and a room with an `items.end` scenery keyword would otherwise emit a
--- file that does not compile.
--- @param s any
--- @return boolean
function M.is_identifier(s)
    if type(s) ~= "string" then return false end
    if KEYWORDS[s] then return false end
    return s:match("^[%a_][%w_]*$") ~= nil
end

--- A string as a double-quoted Lua literal.
---
--- Control bytes become `\ddd` with **three digits always**. Two would be
--- shorter and wrong: `"\0" .. "1"` emitted as `\01` reads back as byte 1
--- followed by nothing, because Lua takes up to three digits and `01` is a
--- complete escape only if the next character is not a digit.
---
--- Bytes at or above 0x80 pass through untouched — that is what UTF-8 prose is,
--- and escaping it would turn every accented character in a description into
--- four unreadable digits.
--- @param s string
--- @return string
function M.quote(s)
    local out = s:gsub("[\\\"]", "\\%0")
    out = out:gsub("\n", "\\n"):gsub("\r", "\\r"):gsub("\t", "\\t")
    out = out:gsub("[%z\1-\8\11\12\14-\31\127]", function(c)
        return string.format("\\%03d", c:byte())
    end)
    return '"' .. out .. '"'
end

--- The smallest long-bracket level that can safely hold `s`, or nil.
---
--- Both `]==]` *and* `[==[` are scanned for. Only the closing form can truncate
--- the literal, but a `[==[` inside would nest and change where it ends, and the
--- cost of skipping a level is one character.
---
--- Returns nil when a long bracket cannot hold the string at all: a `\r` would
--- be normalized by the parser, and a trailing `]` would run into the closing
--- bracket. Those fall back to a quoted literal, which is always correct.
--- @param s string
--- @return string|nil open, string|nil close
function M.long_bracket(s)
    if s:find("\r", 1, true) then return nil end
    if s:sub(-1) == "]" then return nil end

    for level = 0, 8 do
        local eq = string.rep("=", level)
        if not s:find("]" .. eq .. "]", 1, true)
           and not s:find("[" .. eq .. "[", 1, true) then
            return "[" .. eq .. "[", "]" .. eq .. "]"
        end
    end
    return nil
end

--- A number as source that reads back as exactly this number.
---
--- Three traps, all of which produced wrong output before they were handled:
---
--- * **NaN and the infinities** have no literal. `0/0` is not a number token,
---   and emitting `inf` produces a nil global.
--- * **An integral float must keep its point.** On 5.3+ `math.type(2.0)` is
---   "float" and `math.type(2)` is "integer"; emitting `2` for `2.0` silently
---   changes the type, and `speed = 1.0` would come back as an integer.
--- * **`%.17g` is not the answer.** It round-trips every double, and renders
---   every authored `speed = 1.2` as `1.1999999999999999`. Trying increasing
---   precision and stopping at the first that reads back equal gives `1.2` for
---   1.2 and full precision only where it is needed.
--- @param n number
--- @return string|nil source, string|nil err
function M.number(n)
    if not finite(n) then
        return nil, tostring(n) .. " has no Lua literal (NaN and the infinities do not)"
    end

    -- 5.3+ has an integer subtype; LuaJIT and 5.1 do not.
    if math.type then
        if math.type(n) == "integer" then return string.format("%d", n) end
    elseif n % 1 == 0 and math.abs(n) < 2 ^ 53 then
        return string.format("%d", n)
    end

    -- A float. Integral ones need the point kept.
    if n % 1 == 0 and math.abs(n) < 2 ^ 53 then
        return string.format("%.1f", n)
    end

    for _, fmt in ipairs({ "%.14g", "%.15g", "%.16g", "%.17g" }) do
        local text = string.format(fmt, n)
        if tonumber(text) == n then return text end
    end
    return nil, "no representation round-trips for " .. tostring(n)
end

-- ─── Whether a value can be written at all ───────────────────────────────────

local function check(value, depth, path, ancestors)
    local t = type(value)

    if t == "nil" or t == "boolean" or t == "string" then return true end
    if t == "number" then
        local _, err = M.number(value)
        if err then
            return false, (path == "" and "the value" or path) .. ": " .. err
        end
        return true
    end
    if t ~= "table" then
        return false, (path == "" and "the value" or path)
            .. " is a " .. t .. ", which has no source form"
    end

    if depth > M.MAX_DEPTH then
        return false, "nesting deeper than " .. M.MAX_DEPTH .. " at "
            .. (path == "" and "the top level" or path)
    end
    -- On the *ancestor* path, not globally: the same subtable appearing twice
    -- is legal and is written twice, while a table that is its own ancestor has
    -- no bottom.
    if ancestors[value] then
        return false, (path == "" and "the value" or path) .. " contains itself"
    end
    ancestors[value] = true

    for k, v in pairs(value) do
        local kt = type(k)
        if kt ~= "string" and kt ~= "number" then
            return false, join(path, tostring(k)) .. ": a " .. kt
                .. " cannot be a table key in source"
        end
        if kt == "number" and not finite(k) then
            return false, (path == "" and "the value" or path)
                .. " has " .. tostring(k) .. " as a key"
        end
        local ok, err = check(v, depth + 1, join(path, k), ancestors)
        if not ok then
            ancestors[value] = nil
            return false, err
        end
    end

    ancestors[value] = nil
    return true
end

--- Could this value be written as source?
---
--- Same shape as `jsonsafe.check`, and deliberately not the same rules — see the
--- note at the top of this file. Names the offending path, because "cannot
--- serialize" without a location is a bug report you cannot act on.
--- @param value any
--- @return boolean ok, string|nil reason
function M.check(value)
    return check(value, 0, "", {})
end

-- ─── Emitting ────────────────────────────────────────────────────────────────

--- Is this table a pure sequence, `1..n` with nothing else?
local function sequence_length(t)
    local n = 0
    for _ in ipairs(t) do n = n + 1 end
    local count = 0
    for _ in pairs(t) do count = count + 1 end
    return n, count == n
end

--- Default key order: the array part in order, then string keys sorted.
---
--- Sorted rather than `pairs` order, because `pairs` is not stable across runs
--- and a file that reorders itself between writes is a file whose diff is
--- noise. `codegen_d` sorted its exits and not its scenery items, so half its
--- output shuffled.
local function default_order(t)
    local n = sequence_length(t)
    local keys = {}
    for k in pairs(t) do
        if not (type(k) == "number" and k >= 1 and k <= n and k % 1 == 0) then
            keys[#keys + 1] = k
        end
    end
    table.sort(keys, function(a, b) return tostring(a) < tostring(b) end)
    return keys
end

local emit_value

--- One table key, as it appears before the `=`.
local function emit_key(k)
    if M.is_identifier(k) then return k end
    if type(k) == "number" then return "[" .. M.number(k) .. "]" end
    return "[" .. M.quote(k) .. "]"
end

--- A string: long bracket when it is prose with newlines, quoted otherwise.
local function emit_string(s)
    if not s:find("\n", 1, true) then return M.quote(s) end

    local open, close = M.long_bracket(s)
    if not open then return M.quote(s) end

    -- **An opening long bracket eats the newline that immediately follows it.**
    -- That is exactly what the newline written here is for: it is consumed by
    -- the parser, and the content begins at column 0 of the next line, byte for
    -- byte. So a string that itself starts with a newline needs nothing extra —
    -- it keeps its blank first line because the eaten newline is the separator
    -- rather than part of the value. (Adding a second one here "to compensate"
    -- is the obvious move and it silently prepends a blank line on every
    -- round trip.)
    --
    -- Nothing is added at the end either. The closing bracket goes flush against
    -- the last character, so a value not ending in a newline does not acquire
    -- one, and indentation is never inserted before the close — indenting it
    -- puts that indentation *inside* the string, which is how a description
    -- grows four spaces on every read-and-rewrite.
    return open .. "\n" .. s .. close
end

local function emit_table(t, indent, opts, depth, path)
    local n, pure = sequence_length(t)
    local keys = (opts.order and opts.order(t, path)) or default_order(t)

    if n == 0 and #keys == 0 then return "{}" end

    local inner = indent .. opts.indent

    -- A short table of scalars on one line.
    --
    -- `tags = { "indoor", "damp" }` reads as one fact; spread over four lines it
    -- reads as four. The same is true of a map — `damage = { min = 3, max = 7 }`
    -- is one number pair, and three lines make it look like structure.
    do
        local scalar = true
        for _, v in pairs(t) do
            if type(v) == "table" then scalar = false break end
        end
        if scalar then
            local parts = {}
            for i = 1, n do
                parts[#parts + 1] = emit_value(t[i], inner, opts, depth + 1, join(path, i))
            end
            for _, k in ipairs(keys) do
                parts[#parts + 1] = emit_key(k) .. " = "
                    .. emit_value(t[k], inner, opts, depth + 1, join(path, k))
            end
            local oneline = "{ " .. table.concat(parts, ", ") .. " }"
            -- Not when a comment belongs above one of these keys: the whole
            -- point of a section comment is the line break before it.
            local commented = false
            if opts.comment_for then
                for _, k in ipairs(keys) do
                    if opts.comment_for(t, path, k) then commented = true break end
                end
            end
            if not commented
               and #oneline + #indent <= opts.inline_width
               and not oneline:find("\n", 1, true) then
                return oneline
            end
        end
    end
    local _ = pure

    local lines = { "{" }

    for i = 1, n do
        lines[#lines + 1] = inner
            .. emit_value(t[i], inner, opts, depth + 1, join(path, i)) .. ","
    end

    -- `=` aligned within this literal. Deterministic given the value, so it
    -- costs nothing in idempotence; adding a long key reformats this one table,
    -- which is irrelevant when the file is regenerated wholesale anyway.
    local width = 0
    local rendered = {}
    for _, k in ipairs(keys) do
        local text = emit_key(k)
        rendered[#rendered + 1] = { key = k, text = text }
        if #text > width then width = #text end
    end

    for _, entry in ipairs(rendered) do
        local comments = opts.comment_for and opts.comment_for(t, path, entry.key)
        if comments then
            -- A blank line before a section, so the block reads as a block.
            if #lines > 1 then lines[#lines + 1] = "" end
            for _, c in ipairs(comments) do
                lines[#lines + 1] = inner .. "-- " .. c
            end
        end
        local pad = string.rep(" ", width - #entry.text)
        lines[#lines + 1] = inner .. entry.text .. pad .. " = "
            .. emit_value(t[entry.key], inner, opts, depth + 1, join(path, entry.key))
            .. ","
    end

    lines[#lines + 1] = indent .. "}"
    return table.concat(lines, "\n")
end

emit_value = function(value, indent, opts, depth, path)
    local t = type(value)
    if t == "nil" then return "nil" end
    if t == "boolean" then return tostring(value) end
    if t == "number" then return (M.number(value)) end
    if t == "string" then return emit_string(value) end
    return emit_table(value, indent, opts, depth, path)
end

--- Render a value as Lua source that loads back to an equal value.
---
--- @param value any
--- @param opts table|nil {
---     indent       = "    ",
---     order        = function(tbl, path) -> array of keys,
---     comments     = { [key] = { "weapon" } }
---                  | function(tbl, path, key) -> { "weapon" } | nil,
---     inline_width = 78,
--- }
---
--- `comments` takes a function as well as a table for the same reason `order`
--- does: a file holds an *array* of records, and which key deserves a section
--- comment depends on the record. A flat table would put the same comment above
--- the same key in every record, or in none.
--- @return string|nil source, string|nil err
function M.value(value, opts)
    local ok, why = M.check(value)
    if not ok then return nil, why end

    opts = opts or {}

    local comment_for
    if type(opts.comments) == "function" then
        comment_for = opts.comments
    elseif type(opts.comments) == "table" then
        local flat = opts.comments
        comment_for = function(_, _, key) return flat[key] end
    end

    return emit_value(value, "", {
        indent       = opts.indent or "    ",
        order        = opts.order,
        comment_for  = comment_for,
        inline_width = opts.inline_width or M.INLINE_WIDTH,
    }, 0, "")
end

--- A whole module file: header comments, then `return <value>`, newline
--- terminated. What codegen actually calls.
--- @param value any
--- @param opts table|nil  as `M.value`, plus `header` = array of comment lines
--- @return string|nil source, string|nil err
function M.module(value, opts)
    opts = opts or {}
    local body, err = M.value(value, opts)
    if not body then return nil, err end

    local lines = {}
    for _, line in ipairs(opts.header or {}) do
        lines[#lines + 1] = (line == "" and "--" or ("-- " .. line))
    end
    lines[#lines + 1] = "return " .. body
    lines[#lines + 1] = ""
    return table.concat(lines, "\n")
end

return M

File: tests/mudlib/serialize.rs
//! `lib/serialize.lua` — Lua values back into Lua source.
//!
//! Three properties, and they are not the same property:
//!
//! * **P1 idempotence** — `emit(load(emit(v)))` byte-equals `emit(v)`. This is
//!   the one that catches formatting bugs: the indented closing `]]` that put
//!   four spaces inside a description and grew them on every write, the leading
//!   newline an opening long bracket eats, `%.17g` rendering `1.2` as
//!   `1.1999999999999999`.
//! * **P2 fidelity** — `load(emit(v))` deep-equals `v`.
//! * **P3 refusal** — everything it cannot write is refused *by name*, rather
//!   than emitted as something that will not compile. `codegen_d` concatenated
//!   strings, so a room title containing a quote produced a broken file and
//!   nothing said so until the next reload.
//!
//! Through `boot_fixture_with_probe` rather than a stubbed `mlua::Lua`: this is
//! mudlib code and it should be asked what it does inside the engine. It happens
//! to be pure, which is why the cases can be table-driven.

use crate::common::RealVm;

/// Probe source travels as one input line, so these snippets are written
/// readably and folded here. No `--` comments inside them: folded onto one line
/// a comment swallows everything after it.
fn one_line(src: &str) -> String {
    src.lines()
        .map(str::trim)
        .filter(|l| !l.is_empty())
        .collect::<Vec<_>>()
        .join(" ")
}

/// A `deep_equal` and a round-trip harness, defined once in Lua.
///
/// Loading the emitted source with `load` rather than comparing strings is what
/// makes P2 a claim about *values* — a formatting change should not fail it.
/// `roundtrip` then emits a second time and compares bytes, which is P1.
const HARNESS: &str = r#"
    S = require('lib.serialize')
    function deep_equal(a, b)
        if type(a) ~= type(b) then return false end
        if type(a) ~= 'table' then return a == b end
        for k, v in pairs(a) do if not deep_equal(v, b[k]) then return false end end
        for k in pairs(b) do if a[k] == nil then return false end end
        return true
    end
    function roundtrip(value)
        local src, err = S.module(value)
        if not src then return 'refused: ' .. tostring(err) end
        local chunk, load_err = load(src)
        if not chunk then return 'did not compile: ' .. tostring(load_err) .. ' >> ' .. src end
        local ok, back = pcall(chunk)
        if not ok then return 'did not run: ' .. tostring(back) end
        if not deep_equal(value, back) then return 'value changed >> ' .. src end
        local again = S.module(back)
        if again ~= src then
            return 'not idempotent >> FIRST ' .. src .. ' SECOND ' .. again
        end
        return 'ok'
    end
"#;

struct Vm(RealVm);

impl Vm {
    fn new() -> Self {
        let mut vm = RealVm::boot_fixture_with_probe();
        vm.eval(&one_line(&format!("{HARNESS} return 'ready'"))).unwrap();
        Self(vm)
    }

    /// Evaluate a readably-written snippet.
    fn run(&mut self, src: &str) -> String {
        self.0.eval(&one_line(src)).unwrap()
    }
}

/// Every string that has ever broken a naive emitter, in one table so adding a
/// case is one line.
#[test]
fn hostile_strings_round_trip() {
    let mut vm = Vm::new();

    let out = vm.run(r#"
            local cases = {
                'plain',
                '',
                'a "quoted" title',
                "an apostrophe's place",
                'a backslash \\ here',
                'tab\there',
                'newline\nhere',
                'crlf\r\nhere',
                'trailing newline\n',
                '\nleading newline',
                '\n\ntwo leading newlines',
                'closing ]] bracket',
                'level one ]=] bracket',
                'level two ]==] bracket',
                'opening [[ bracket',
                'both [[ and ]] brackets',
                'ends with a bracket ]',
                'multi\nline with ]] inside',
                'nul\0byte',
                'byte \127 delete',
                'caf\195\169',
                'em \226\128\148 dash',
                'a percent %s format',
                'a {colour} tag',
            }
            local bad = {}
            for i, s in ipairs(cases) do
                local r = roundtrip({ value = s })
                if r ~= 'ok' then bad[#bad+1] = i .. ': ' .. r end
            end
            return #bad == 0 and 'ok' or table.concat(bad, '\n')
        "#,
        );
    assert_eq!(out, "ok");
}

/// The float trap. `%.17g` round-trips everything and renders every authored
/// `speed = 1.2` as `1.1999999999999999`; `%d` for an integral float silently
/// changes `math.type`.
#[test]
fn numbers_round_trip_without_growing_digits() {
    let mut vm = Vm::new();

    let out = vm.run(r#"
            local cases = { 0, 1, -1, 42, -42, 0.5, 1.2, 0.1, 2.0, -0.0,
                            1e15, 1e-15, math.pi, 2^53, -2^53, 1/3 }
            local bad = {}
            for i, n in ipairs(cases) do
                local r = roundtrip({ value = n })
                if r ~= 'ok' then bad[#bad+1] = i .. ' (' .. tostring(n) .. '): ' .. r end
            end
            return #bad == 0 and 'ok' or table.concat(bad, '\n')
        "#,
        );
    assert_eq!(out, "ok");

    // The specific regression: a value a builder typed should come back looking
    // like the value a builder typed.
    assert_eq!(vm.run("return S.number(1.2)"), "1.2");
    assert_eq!(vm.run("return S.number(0.1)"), "0.1");
    assert_eq!(vm.run("return S.number(42)"), "42");
}

/// `false` must survive. It is falsy, so every `if v then emit(v) end` drops it,
/// and a mob whose `aggressive` quietly became nil is a mob that stops fighting.
#[test]
fn false_and_empty_survive() {
    let mut vm = Vm::new();

    assert_eq!(
        vm.run("return roundtrip({ two_handed = false, aggressive = false })"),
        "ok"
    );
    assert_eq!(vm.run("return roundtrip({ tags = {}, exits = {} })"), "ok");

    let src = vm.run("return S.value({ two_handed = false })");
    assert!(src.contains("two_handed = false"), "{src}");
}

/// Keys: identifiers bare, everything else bracketed, keywords never bare.
#[test]
fn keys_are_emitted_in_a_form_that_compiles() {
    let mut vm = Vm::new();

    assert_eq!(
        vm.run("return roundtrip({ plain = 1, ['with space'] = 2, ['end'] = 3, \
                                ['1numeric'] = 4, [5] = 'five', ['a-dash'] = 6 })"
        ),
        "ok"
    );

    // `end = 1` is a syntax error, and a room with an `end` scenery keyword
    // would otherwise emit a file that does not compile.
    assert_eq!(vm.run("return tostring(S.is_identifier('end'))"), "false");
    assert_eq!(vm.run("return tostring(S.is_identifier('short'))"), "true");
    assert_eq!(vm.run("return tostring(S.is_identifier('with space'))"), "false");
}

/// A table that is both a list and a map is fine here, though `jsonsafe` refuses
/// one. That difference is the reason these are two modules.
#[test]
fn a_mixed_list_and_map_is_written() {
    let mut vm = Vm::new();

    assert_eq!(
        vm.run("return roundtrip({ 'a', 'b', name = 'both', count = 2 })"),
        "ok"
    );
    // …and `jsonsafe` still refuses it, so neither has quietly adopted the
    // other's rules.
    assert_eq!(
        vm.run("return tostring((require('lib.jsonsafe').check({ 'a', name = 'x' })))"),
        "false"
    );
}

/// Everything it cannot write is refused, and the reason names the path.
///
/// "Cannot serialize" without a location is a bug report you cannot act on, and
/// the whole point of refusing is that a builder can go and fix the field.
#[test]
fn what_cannot_be_written_is_refused_by_name() {
    let mut vm = Vm::new();

    let cases = [
        ("{ a = { b = function() end } }", "a.b", "function"),
        ("{ nan = 0/0 }", "nan", "NaN"),
        ("{ inf = math.huge }", "inf", "inf"),
        ("{ co = coroutine.create(function() end) }", "co", "thread"),
    ];
    for (expr, path, needle) in cases {
        let out = vm.run(&format!(
                "local ok, why = S.check({expr}) return tostring(ok) .. '|' .. tostring(why)"
            ));
        assert!(out.starts_with("false|"), "{expr} should be refused: {out}");
        assert!(out.contains(path), "the reason should name '{path}': {out}");
        assert!(
            out.to_lowercase().contains(&needle.to_lowercase()),
            "the reason should say what was wrong ({needle}): {out}"
        );
    }

    // A cycle, named. Detected on the *ancestor* path, so the same subtable
    // appearing twice is still legal.
    let out = vm.run("local t = { name = 'x' } t.self = t \
             local ok, why = S.check(t) return tostring(ok) .. '|' .. tostring(why)",
        );
    assert!(out.starts_with("false|"), "{out}");
    assert!(out.contains("contains itself"), "{out}");

    let shared = vm.run("local leaf = { 1, 2 } \
             return tostring((S.check({ a = leaf, b = leaf })))",
        );
    assert_eq!(shared, "true", "the same table twice is not a cycle");

    // And `M.value` refuses rather than emitting something broken.
    let out = vm.run("local src, err = S.value({ f = print }) return tostring(src) .. '|' .. tostring(err)");
    assert!(out.starts_with("nil|"), "{out}");
}

/// Prose uses a long bracket, at a level that does not collide with its content.
#[test]
fn multi_line_prose_uses_a_long_bracket_that_does_not_truncate() {
    let mut vm = Vm::new();

    let src = vm.run("return S.value({ description = 'Line one.\\nLine two.' })");
    assert!(src.contains("[["), "prose should use a long bracket:\n{src}");

    // A `]]` in the content forces a level up rather than truncating the string.
    let src = vm.run("return S.value({ description = 'Has ]] inside.\\nAnd more.' })");
    assert!(src.contains("[=["), "should step up a level:\n{src}");
    assert_eq!(
        vm.run("return roundtrip({ description = 'Has ]] inside.\\nAnd more.' })"),
        "ok"
    );

    // **Nothing is inserted around the content.** The old emitter wrote the
    // closing `]]` indented to match the field, which put that indentation
    // *inside* the string — so a description grew four spaces on every
    // read-and-rewrite, for ever. Nested two levels deep here, where the
    // temptation to indent is strongest.
    let src = vm.run("return S.value({ nested = { description = 'a\\nb\\n' } })");
    assert!(
        src.contains("\n]]") && !src.contains("    ]]"),
        "the closing bracket must not be indented:\n{src}"
    );
    // A string not ending in a newline closes flush against its last character,
    // for the same reason: a newline there would be a newline in the value.
    let src = vm.run("return S.value({ nested = { description = 'a\\nb' } })");
    assert!(src.contains("b]]"), "no newline may be invented before the close:\n{src}");

    // And both survive the trip, which is the property the formatting serves.
    assert_eq!(vm.run("return roundtrip({ d = 'a\\nb\\n' })"), "ok");
    assert_eq!(vm.run("return roundtrip({ d = 'a\\nb' })"), "ok");
}

/// Ordering is stable, so a file that has not changed produces no diff.
///
/// `codegen_d` sorted its exits and iterated its scenery items with `pairs`, so
/// half of every generated room reshuffled between writes and every diff was
/// noise.
#[test]
fn key_order_is_deterministic() {
    let mut vm = Vm::new();

    let once = vm.run("return S.value({ zebra = 1, apple = 2, mango = 3, ['_x'] = 4 })");
    for _ in 0..5 {
        let again = vm.run("return S.value({ zebra = 1, apple = 2, mango = 3, ['_x'] = 4 })");
        assert_eq!(once, again, "emission is not stable across calls");
    }
    // Sorted, so the order is predictable rather than merely repeatable.
    let apple = once.find("apple");
    let zebra = once.find("zebra");
    assert!(apple < zebra, "string keys should be sorted:\n{once}");
}

/// An explicit order wins, which is how codegen puts schema fields in schema
/// order rather than alphabetically.
#[test]
fn a_caller_can_impose_its_own_order() {
    let mut vm = Vm::new();

    let src = vm.run("return S.value({ description = 'd', id = 'x', short = 's' }, { \
                order = function() return { 'id', 'short', 'description' } end })",
        );
    let id = src.find("id ");
    let short = src.find("short ");
    let desc = src.find("description ");
    assert!(id < short && short < desc, "declared order should win:\n{src}");
}

/// A short array of scalars stays on one line. `tags = { "indoor", "damp" }` is
/// one fact; spread over four lines it reads as four.
#[test]
fn short_scalar_arrays_are_inline() {
    let mut vm = Vm::new();

    let src = vm.run("return S.value({ tags = { 'indoor', 'damp' } })");
    assert!(
        src.contains(r#"{ "indoor", "damp" }"#),
        "a short array should be inline:\n{src}"
    );

    // A long one is not, and still round-trips.
    assert_eq!(
        vm.run("local t = {} for i = 1, 40 do t[i] = 'tag_number_' .. i end \
             return roundtrip({ tags = t })"
        ),
        "ok"
    );
}

/// `M.module` produces a file, header comments and all.
#[test]
fn module_emits_a_loadable_file() {
    let mut vm = Vm::new();

    let src = vm.run("return S.module({ { id = 'crypt.hall' } }, \
                { header = { 'Generated by OLC.', '', 'DO NOT EDIT.' } })",
        );

    assert!(src.starts_with("-- Generated by OLC.\n--\n-- DO NOT EDIT.\n"), "{src}");
    assert!(src.contains("return {"), "{src}");
    assert!(src.ends_with("\n"), "a file should end with a newline");

    assert_eq!(
        vm.run("local src = S.module({ { id = 'crypt.hall' } }) \
             local t = load(src)() return t[1].id"
        ),
        "crypt.hall"
    );
}

/// The realistic shape: a room with every field type at once.
#[test]
fn a_whole_room_round_trips() {
    let mut vm = Vm::new();

    assert_eq!(
        vm.run(r#"return roundtrip({
                {
                    id          = 'crypt.hall',
                    short       = 'The "Great" Hall',
                    description = 'Bone stacked to the vault.\nFour hundred years of it.',
                    light       = 1,
                    smell       = 'Silt, and something older.',
                    tags        = { 'indoor', 'damp' },
                    exits       = { north = 'crypt.stair', down = 'crypt.ossuary' },
                    items       = { ['tide mark'] = 'A brown line, dead level.' },
                    stats       = { corruption = 3 },
                },
            })"#
        ),
        "ok"
    );
}
